File: src/format/comment-rewrite.ts
// Port of google-java-format core/.../java/JavaCommentsHelper.java (the
// `rewrite` path). The outer comment-normalization layer that runs at write time
// on every comment, given the column it starts at. For a javadoc comment it first
// runs the full reflow (formatter.ts); then it re-indents/wraps the lines.

import { formatJavadoc } from "./javadoc/formatter.ts";

const MAX_LINE_LENGTH = 100;

/**
 * Rewrite a comment for output at `column0`. `isLine` is true for `//` comments.
 * Mirrors `JavaCommentsHelper.rewrite`.
 */
export function rewriteComment(text: string, column0: number, isLine: boolean): string {
  if (text.startsWith("/**")) {
    text = formatJavadoc(text, column0);
  }
  const lines = text.split("\n").map(l => (isLine ? l.trim() : trimTrailing(l)));
  if (isLine) return indentLineComments(lines, column0);
  const pc = reformatParamComment(text);
  if (pc !== null) return pc;
  return javadocShaped(lines) ? indentJavadoc(lines, column0) : preserveIndentation(lines, column0);
}

// gjf's PARAMETER_COMMENT: `/*name=*/` -> `/* name= */` (identifier, optional
// varargs `...`). Returns null when the text is not a parameter comment.
const PARAM_COMMENT = /^\/\*\s*([A-Za-z_$][\w$]*(?:\.\.\.)?)\s*=\s*\*\/$/;

export function reformatParamComment(text: string): string | null {
  const m = text.match(PARAM_COMMENT);
  return m ? `/* ${m[1]}= */` : null;
}

// For a non-javadoc-shaped block comment, shift the block to `column0` but keep
// relative indentation.
function preserveIndentation(lines: string[], column0: number): string {
  let startCol = -1;
  for (let i = 1; i < lines.length; i++) {
    const idx = lines[i].search(/\S/);
    if (idx >= 0 && (startCol === -1 || idx < startCol)) startCol = idx;
  }
  let out = lines[0];
  const pad = " ".repeat(column0);
  for (let i = 1; i < lines.length; i++) {
    out += "\n" + pad;
    out += startCol >= 0 && lines[i].length >= startCol ? lines[i].slice(startCol) : lines[i];
  }
  return out;
}

// Re-indent a javadoc-shaped block comment: first line as-is, continuations at
// `column0+1` with a `*` prefix.
function indentJavadoc(lines: string[], column0: number): string {
  let out = lines[0].trim();
  const pad = " ".repeat(column0 + 1);
  for (let i = 1; i < lines.length; i++) {
    out += "\n" + pad;
    let line = lines[i].trim();
    if (!line.startsWith("*")) {
      out += "* ";
    }
    out += line;
  }
  return out;
}

// Wrap and re-indent line comments.
function indentLineComments(lines: string[], column0: number): string {
  lines = wrapLineComments(lines, column0);
  let out = lines[0].trim();
  const pad = " ".repeat(column0);
  for (let i = 1; i < lines.length; i++) out += "\n" + pad + lines[i].trim();
  return out;
}

// Preserve special `//noinspection` / `//$NON-NLS-x$` comments (no leading space).
const MISSING_SPACE_PREFIX = /^(\/\/+)(?!noinspection|\$NON-NLS-\d+\$)[^\s/]/;

function wrapLineComments(lines: string[], column0: number): string[] {
  const result: string[] = [];
  for (let line of lines) {
    const m = line.match(MISSING_SPACE_PREFIX);
    if (m) {
      const length = m[1].length;
      line = "/".repeat(length) + " " + line.slice(length);
    }
    if (line.startsWith("// MOE:")) {
      result.push(line);
      continue;
    }
    while (line.length + column0 > MAX_LINE_LENGTH) {
      let idx = MAX_LINE_LENGTH - column0;
      while (idx >= 2 && !/\s/.test(line[idx])) idx--;
      if (idx <= 2) break;
      result.push(line.slice(0, idx));
      line = "//" + line.slice(idx);
    }
    result.push(line);
  }
  return result;
}

function javadocShaped(lines: string[]): boolean {
  if (lines.length === 0) return false;
  const first = lines[0].trim();
  if (first.startsWith("/**")) return true;
  if (!first.startsWith("/*")) return false;
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i].trim().startsWith("*")) return false;
  }
  return true;
}

function trimTrailing(s: string): string {
  return s.replace(/\s+$/, "");
}

File: togo/internal/format/comment_rewrite.go
// Port of src/format/comment-rewrite.ts (gjf's JavaCommentsHelper.rewrite path).
//
// Outer comment normalization, run at write time on every comment given the
// column it starts at. Javadoc first goes through the full reflow engine.

package format

import (
	"regexp"
	"strings"
	"unicode/utf16"

	"github.com/nikeee/cappu/internal/format/javadoc"
)

const commentMaxLineLength = 100

// rewriteComment rewrites a comment for output at column0. isLine is true for
// `//` comments. Mirrors JavaCommentsHelper.rewrite.
func rewriteComment(text string, column0 int, isLine bool) string {
	if strings.HasPrefix(text, "/**") {
		text = javadoc.FormatJavadoc(text, column0)
	}
	rawLines := strings.Split(text, "\n")
	lines := make([]string, len(rawLines))
	for i, l := range rawLines {
		if isLine {
			lines[i] = strings.TrimSpace(l)
		} else {
			lines[i] = strings.TrimRight(l, " \t\r\n\f\v")
		}
	}
	if isLine {
		return indentLineComments(lines, column0)
	}
	if pc, ok := reformatParamComment(text); ok {
		return pc
	}
	var out string
	if javadocShaped(lines) {
		out = indentJavadoc(lines, column0)
	} else {
		out = preserveIndentation(lines, column0)
	}
	// Re-indenting pads every continuation line to column0, including the blank
	// ones. gjf's writer drops that trailing run when it emits the line; our doc
	// writer only trims at breaks it emits itself, so do it here.
	return trimLineEnds(out)
}

// trimLineEnds strips trailing spaces and tabs from every line.
func trimLineEnds(s string) string {
	if !strings.ContainsAny(s, " \t") {
		return s
	}
	lines := strings.Split(s, "\n")
	for i, l := range lines {
		lines[i] = strings.TrimRight(l, " \t")
	}
	return strings.Join(lines, "\n")
}

var paramCommentRe = regexp.MustCompile(`^/\*\s*([\p{L}_$][\p{L}\p{N}_$]*(?:\.\.\.)?)\s*=\s*\*/$`)

// reformatParamComment normalizes `/*name=*/` -> `/* name= */`; ok=false if not
// a parameter comment.
func reformatParamComment(text string) (string, bool) {
	m := paramCommentRe.FindStringSubmatch(text)
	if m == nil {
		return "", false
	}
	return "/* " + m[1] + "= */", true
}

func preserveIndentation(lines []string, column0 int) string {
	startCol := -1
	for i := 1; i < len(lines); i++ {
		idx := strings.IndexFunc(lines[i], func(r rune) bool { return !isSpaceRune(r) })
		if idx >= 0 && (startCol == -1 || idx < startCol) {
			startCol = idx
		}
	}
	var b strings.Builder
	b.WriteString(lines[0])
	pad := strings.Repeat(" ", column0)
	for i := 1; i < len(lines); i++ {
		b.WriteString("\n")
		b.WriteString(pad)
		if startCol >= 0 && len(lines[i]) >= startCol {
			b.WriteString(lines[i][startCol:])
		} else {
			b.WriteString(lines[i])
		}
	}
	return b.String()
}

func indentJavadoc(lines []string, column0 int) string {
	var b strings.Builder
	b.WriteString(strings.TrimSpace(lines[0]))
	pad := strings.Repeat(" ", column0+1)
	for i := 1; i < len(lines); i++ {
		b.WriteString("\n")
		b.WriteString(pad)
		line := strings.TrimSpace(lines[i])
		if !strings.HasPrefix(line, "*") {
			b.WriteString("* ")
		}
		b.WriteString(line)
	}
	return b.String()
}

func indentLineComments(lines []string, column0 int) string {
	lines = wrapLineComments(lines, column0)
	var b strings.Builder
	b.WriteString(strings.TrimSpace(lines[0]))
	pad := strings.Repeat(" ", column0)
	for i := 1; i < len(lines); i++ {
		b.WriteString("\n")
		b.WriteString(pad)
		b.WriteString(strings.TrimSpace(lines[i]))
	}
	return b.String()
}

var missingSpacePrefix = regexp.MustCompile(`^(//+)(?:[^\s/])`)
var allowedNoSpace = regexp.MustCompile(`^//(?:noinspection|\$NON-NLS-\d+\$)`)

func wrapLineComments(lines []string, column0 int) []string {
	var result []string
	for _, line := range lines {
		if m := missingSpacePrefix.FindStringSubmatch(line); m != nil && !allowedNoSpace.MatchString(line) {
			length := len(m[1])
			line = strings.Repeat("/", length) + " " + line[length:]
		}
		if strings.HasPrefix(line, "// MOE:") {
			result = append(result, line)
			continue
		}
		// Column arithmetic runs in UTF-16 code units (the TypeScript build's
		// string indices, gjf's Java chars): counting bytes made every comment
		// holding a non-ASCII character wrap a few columns early.
		units := utf16.Encode([]rune(line))
		for len(units)+column0 > commentMaxLineLength {
			idx := commentMaxLineLength - column0
			for idx >= 2 && !isSpaceUnit(units[idx]) {
				idx--
			}
			if idx <= 2 {
				break
			}
			result = append(result, string(utf16.Decode(units[:idx])))
			units = append(utf16.Encode([]rune("//")), units[idx:]...)
		}
		result = append(result, string(utf16.Decode(units)))
	}
	return result
}

func javadocShaped(lines []string) bool {
	if len(lines) == 0 {
		return false
	}
	first := strings.TrimSpace(lines[0])
	if strings.HasPrefix(first, "/**") {
		return true
	}
	if !strings.HasPrefix(first, "/*") {
		return false
	}
	for i := 1; i < len(lines); i++ {
		if !strings.HasPrefix(strings.TrimSpace(lines[i]), "*") {
			return false
		}
	}
	return true
}

func isSpaceRune(r rune) bool {
	return r == ' ' || r == '\t' || r == '\n' || r == '\r' || r == '\f' || r == '\v'
}
func isSpaceUnit(u uint16) bool {
	return u == ' ' || u == '\t' || u == '\n' || u == '\r' || u == '\f' || u == '\v'
}
